### DISCUSSION TITLE
[Feature] some type of control over albums in the storage template engine

### DISCUSSION THREAD

**Discussion by mmomjian** [immich-app/immich#6810] [Feature Request]
### The feature

I have a number of photos that are in multiple albums, and I would like to have some control over how the storage engine stores these files as I use the `{{album}}` template. I believe it currently select a random album and names it based on that album name.

I think there are a few ways to implement this, the easiest option may be in `server/src/domain/storage-template/storage-template.service.ts:191`, we could change `albumName = albums?.[0]?.albumName || null;` to `albumName = albums?.sort().[0]?.albumName || null;`. I am not well versed in TS so the syntax may be incorrect but essentially this would sort by album in alphabetical order, allowing me to name "hidden" albums for example `zAlbum Name`.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by mmomjian** [comment]
Update: based on some testing it looks like Immich currently uses the most recent album for the storage template engine. I do think some better control over this would be nice, but for now I will make a documentation note of this fact.

**Comment by nthamling** [comment]
I fully agree with this idea and would also like to have some control over how the storage engine stores these files. I also use the {{album}} template. 

In my case, I have all my photos from the last 20 years sorted into albums according to events (holidays, birthdays, parties, ...). This is my primary sorting method. The photos should be organised by event in folders. With this method, I have survived several changes in the storage of my photos over the last few years in a reasonably structured way (from self-hosted solutions to cloud services and back). That's why I want to keep it that way.

I recently created a "best off" album for my daughter's 18th birthday. Unfortunately, the "STORAGE TEMPLATE MIGRATION" job moved all the selected images from the event albums to the "Best-Off" album. But that doesn't fit in with my filing strategy...

So it would definitely be cool if immich had more control over which folder the job saves the images in. 

I wonder what is the best way to do this?
- Tick the box for albums to indicate whether the job should take them into account when sorting.
- Organise the albums by priority. If a photo is in several albums, it will be stored in the album with the highest priority.
- ???

P.S.: I can't code. But I love Immich. Immich has helped me put all my family photos from the last 20 years from different places and different structures into a good structure. It's beautiful and most things work as expected - especially the face recognition is much better than I've ever experienced with other programmes.

**Comment by radh21301** [comment]
This would definitely be an awesome thing to have.

I currently run '[Immich-Frame](https://github.com/3rob3/ImmichFrame)' - an amazing app that turns a laptop/tablet etc. into a beautiful picture frame using the images from immich! It has a feature where the recently displayed pictures (recent 50 or something) are added to an album (we decide which album that app needs to use) -> This is useful if I want to do something with those images (delete/archive/edit date/edit location etc). What this does is move my pictures around, from an album where I want it ->TO-> immich-frame album. Sort of annoying and confusing the first time I was looking for those images.

**Comment by caeci11ius** [comment]
There are some nice mockups by @szeraax and suggestions as to how this might work using a "primary album" concept in #10839

**Comment by desmondgc** [comment]
A feature like _primary album_ or some other concept allowing granular control over photos belonging to multiple albums would be great. Until then, I wonder if simply reversing the current behaviour and using the **oldest** album would be a more sane default.

My usual workflow is to create event-based albums like "Tom's birthday" and "Wedding anniversary" which I want to be reflected in the file system. It's usually _afterwards_ that I find myself creating additional secondary albums like "Best of 2024" or "Birthday slideshow" or other variations for sharing and/or aggregating.

**Comment by DalaraAFL** [comment]
Maybe im missing something, I am new here and havent looked through the code, but could one solution not be creating hardlinks in all the album folders ? That way the image is just present in all of the albums, but only actually stored once on disk? This might be more user friendly while also reducing UI work required to support albums.

**Comment by ChilledSlim** [comment]
Primary album is necessary.
    /2025-01-05 ABC Concert/ could be an album.  It's an event.  A grouping.
If I then go and add a folder for a person within it (ie: photos of Karen or Concerts), being able to select the most appropriate grouping is useful.

**Comment by nikomiko** [comment]
Currently in the process of migrating my 80k family pictures to Immich. The storage template is a great feature, and should be limited to using a "Main Album" in case a photo were to be placed in different albums. 
I think that Albums may be used in 2 major different ways, considering a personnal usage : 
- grouping pictures based on the context it's been shot in : wedding, trip, family event, etc.
- grouping pictures based on why / where / whom with / etc. you want to look at the pictures : best of XXXX, favorite cars shots, etc.

Immich has already some great automatic grouping features like faces, places, etc. 
But users should be able to decide, based on these info, on a deterministic way to physically sort the pictures.

**Comment by FraYoshi** [comment]
Found myself in need of this right today.. I keep 2 directory trees for my album, one for the ones I am okay being backupped online, and others that I need to stay local. Inside of these I subdivide them only by year, and a couple of nested directories.

Wanted to replicate the same thing on Immich using the templates, but in here, I also create albums to share with the family. Now, the sorting works fine if I add the media to the 2 albums, local and online, but the moment I create a new one to share, these media get moved into the newer album...
My actual solution is to create an album every now and then > and so select all > add to the newer album > delete the old one... but this isn't how it should need to be done.

My proposal aligns with the ones above, which is to add an IF statement where we can define that if the medium is inside an album named "A", then this goes into "A", if "B" goes in "B", else it goes on the root, or into "C".

Another solution I would like, is to do the same thing, but with tags, so if the medium has a specific tag assigned, the directory with the given tag is where all the media having such tag go.

An interesting alternative for BTRFS folks, would be to have reflinks between the albums, and create a directory for each and every album if wanted, with the media in all the places. This would not need exponential space, as it would all be metadata. Also hardlinks for other filesystems might work.

The solution to mark some master album on the web interface is also a sound solution.

Well, I guess I made my point, such a feature would be very welcomed!

**Reply by nadir-ishiguro** [reply]
Had the exact same thought in regards to btrfs/zfs/xfs (all support reflinks). That would be my ideal solution.
